crates/wasmtime/src/compile.rs

```diff
@@ -419,6 +422,28 @@ impl<'a> CompileInputs<'a> {
         ret
     }
 
+    fn clean_symbol(name: String) -> String {
+        // Just to be on the safe side, avoid passing user-provided data to tools
+        // like "perf" or "objdump", and filter the name.
+        let mut cleaned = String::with_capacity(name.len());
+        for c in name.chars().map(|c| {
+            if c.is_alphanumeric() || r"<>[]_-:@$".contains(c) {
+                c
+            } else {
+                '?'
+            }
+        }) {
+            cleaned.push(c);
+        }
```

Comment:
I think you can do something better with iterators, and also we'll want to avoid allocating a string in the common case -- perhaps something like:

- Define a predicate `fn bad_char(c: char) -> bool { ... }` inside this function
- Guard on no bad chars, to allow the common case without allocations; if any bad chars, do an iterator stream filter

```
let name = if name.chars().any(|c| bad_char(c)) {
  Cow::Owned(name.chars().map(|c| if bad_char(c) { '?' } else { c }).collect::<String>())
} else {
  Cow::Borrowed(&name[..])
}
```

Thoughts?

Reply:
Yup, right after I pushed this I found about the `map()` method.  Let me use `Cow` too, as this should reduce quite a bit of writes in the happy path.

Reply:
Done, thanks!